Userland/Libraries/LibCore/FileWatcher.h

```diff
@@ -13,48 +13,114 @@
+class FileWatcherBase {
+public:
+    virtual ~FileWatcherBase() { }
+
+    virtual String class_name() const = 0;
+
+    Result<bool, String> add_watch(String path, InodeWatcherEvent::Type event_mask);
+    Result<bool, String> remove_watch(String path);
+    bool is_watching(String path) { return m_path_to_wd.find(path) != m_path_to_wd.end(); }
+
+protected:
+    FileWatcherBase(int watcher_fd)
+        : m_watcher_fd(watcher_fd)
+    {
+    }
+
+    int m_watcher_fd { -1 };
+    HashMap<String, unsigned> m_path_to_wd;
+    HashMap<unsigned, String> m_wd_to_path;
+};
+
+class BlockingFileWatcher final : public FileWatcherBase {
     AK_MAKE_NONCOPYABLE(BlockingFileWatcher);
 
 public:
-    explicit BlockingFileWatcher(const String& path);
+    explicit BlockingFileWatcher(InodeWatcherFlags = InodeWatcherFlags::None);
     ~BlockingFileWatcher();
+    virtual String class_name() const override { return "BlockingFileWatcher"; }
 
     Optional<FileWatcherEvent> wait_for_event();
-
-private:
-    String m_path;
-    int m_watcher_fd { -1 };
 };
 
-class FileWatcher : public RefCounted<FileWatcher> {
+class FileWatcher final : public FileWatcherBase
+    , public RefCounted<FileWatcher> {
     AK_MAKE_NONCOPYABLE(FileWatcher);
 
 public:
-    static Result<NonnullRefPtr<FileWatcher>, String> watch(const String& path);
+    static Result<NonnullRefPtr<FileWatcher>, String> create(InodeWatcherFlags = InodeWatcherFlags::None);
     ~FileWatcher();
 
+    virtual String class_name() const override { return "FileWatcher"; }
+
     Function<void(FileWatcherEvent)> on_change;
 
 private:
-    FileWatcher(NonnullRefPtr<Notifier>, const String& path);
+    FileWatcher(int watcher_fd, NonnullRefPtr<Notifier>);
 
     NonnullRefPtr<Notifier> m_notifier;
-    String m_path;
+};
+
+}
+
+namespace AK {
+
+template<>
+struct Formatter<Core::FileWatcherEvent> : Formatter<FormatString> {
+    void format(FormatBuilder& builder, const Core::FileWatcherEvent& value)
+    {
+        Formatter<FormatString>::format(builder, "FileWatcherEvent(\"{}\", {})", value.event_path, value.type);
+    }
+};
+
+template<>
+struct Formatter<Core::FileWatcherEvent::Type> : Formatter<FormatString> {
+    void format(FormatBuilder& builder, const Core::FileWatcherEvent::Type& value)
+    {
+        char const* type;
+        switch (value) {
+        case Core::FileWatcherEvent::Type::ChildCreated:
+            type = "ChildCreated";
+            break;
+        case Core::FileWatcherEvent::Type::ChildDeleted:
+            type = "ChildDeleted";
+            break;
+        case Core::FileWatcherEvent::Type::Deleted:
+            type = "Deleted";
+            break;
+        case Core::FileWatcherEvent::Type::ContentModified:
+            type = "ContentModified";
+            break;
+        case Core::FileWatcherEvent::Type::MetadataModified:
+            type = "MetadataModified";
+            break;
+        default:
+            builder.put_string("???");
```

Comment:
Maybe it would be nicer to just `VERIFY_NOT_REACHED()` here instead of trying to truck along?